arrow-libs/core/arrow-core-retrofit/src/main/kotlin/arrow/retrofit/adapter/either/networkhandling/NetworkEitherCallAdapter.kt

```diff
@@ -0,0 +1,87 @@
+package arrow.retrofit.adapter.either.networkhandling
+
+import arrow.core.Either
+import arrow.core.left
+import arrow.core.right
+import okhttp3.Request
+import okio.Timeout
+import retrofit2.Call
+import retrofit2.CallAdapter
+import retrofit2.Callback
+import retrofit2.Response
+import java.io.IOException
+import java.lang.reflect.Type
+
+internal class NetworkEitherCallAdapter<R>(
+  private val successType: Type
+) : CallAdapter<R, Call<Either<CallError, R?>>> {
+
+  override fun adapt(call: Call<R?>): Call<Either<CallError, R?>> = EitherCall(call, successType)
+
+  override fun responseType(): Type = successType
+}
+
+private class EitherCall<R>(
+  private val delegate: Call<R>,
+  private val successType: Type
+) : Call<Either<CallError, R>> {
+
+  override fun enqueue(callback: Callback<Either<CallError, R>>) = delegate.enqueue(
+    object : Callback<R> {
+
+      override fun onResponse(call: Call<R>, response: Response<R>) {
+        callback.onResponse(this@EitherCall, Response.success(response.toEither()))
+      }
+
+      private fun Response<R>.toEither(): Either<CallError, R> {
+        // Http error response (4xx - 5xx)
+        if (!isSuccessful) {
+          val errorBody = errorBody()?.string() ?: ""
+          return HttpError(code(), errorBody).left()
```

Comment:
this should probably use `requireNotNull` or `!!`.

AFIK it's not possible to construct an error response without a body: https://github.com/square/retrofit/blob/master/retrofit/src/main/java/retrofit2/Response.java

Reply:
Ok, I see that the ` error(ResponseBody body, okhttp3.Response rawResponse)` factory method indeed uses `requireNonNull(body)`.
What wonders me is why then the Retrofit public API marks the `errorBody()` as nullable.

Reply:
That makes sense